### PR TITLE
Branching strategy ADR

### PR DESCRIPTION
**Pull request by SergeyMenshykh**
This ADR outlines industry-adopted branching strategies, branching strategies used in SK, and potential strategies that can be considered for SK.

### CODE REVIEW

File: docs/decisions/0030-branching-strategy.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: SergeyMenshykh
+date: 2024-01-04
+deciders: markwallace-microsoft
+consulted:
```
Comment (RogerBarreto) [inline docs/decisions/0030-branching-strategy.md:7]:
```suggestion
consulted: rogerbarreto, dmytrostruk
```

**Review by RogerBarreto — APPROVED** [review]

**Review by moonbox3 — APPROVED** [review]
Great overview of the various strategies considered and their pros/cons.

**Review by juliomenendez — APPROVED** [review]

File: docs/decisions/0030-branching-strategy.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+The Java version of SK adheres to the Git-Flow strategy by being developed in a dedicated development branch. Topic branches are created from the development branch and merged back through pull requests after unit tests and integration test runs. Release branches are also created from the development branch and merged to both the development branch and the 'main' one when a release is considered production-ready. This strategy deviates slightly from vanilla Git-Flow in that release artifacts are generated from release branches rather than from the 'main' branch.
+
+## Decision Drivers  
+- The strategy should be easy to implement and maintain without requiring significant investments.
+- The strategy should allow for maintaining several releases in parallel if required.
+- Ideally, the strategy is intuitive and simple so that everyone familiar with Git can adopt and follow it.
+- Ideally, all SK languages are able to adopt and use the same branching strategy.
```
Comment (markwallace-microsoft) [inline docs/decisions/0030-branching-strategy.md:57]: Some additional drivers:

- Ability to continually deploy new release with minimal overhead
- Ability to release language versions independently and on different schedules
- Allow the .Net, Java and Python teams to be able to operate independantly
- Ability to patch a release (for all languages)
- Consolidation of PR's and Issues to simplify the triage and review process

Also is there a section which explains the current strategy (and associated issues) being used by the Java team?

Comment (SergeyMenshykh) [reply docs/decisions/0030-branching-strategy.md:57]: Thanks, added.

Java strategy is explained in the "SK branching strategies" section:
image
The problem with squash merges is mentioned in the "Branch per SK language" option.

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**